src/asl.ts

```diff
@@ -715,19 +733,57 @@ export class ASL {
               ResultPath: tempArrayPath,
             })!;
 
-        const initializerName = isIdentifier(stmt.initializer)
-          ? stmt.initializer.name
-          : isVariableDecl(stmt.initializer) &&
-            isIdentifier(stmt.initializer.name)
-          ? stmt.initializer.name.name
-          : undefined;
+        const initializer: ASLGraph.SubState | ASLGraph.NodeState = (() => {
+          /**ForInStmt
+           * Assign the value to $.0__[variableName].
+           * Assign the index to the variable decl. If the variable decl is an identifier, it may be carried beyond the ForIn.
+           */
+          if (isForInStmt(stmt)) {
+            const initializerName = isIdentifier(stmt.initializer)
+              ? stmt.initializer.name
+              : isVariableDecl(stmt.initializer) &&
+                isIdentifier(stmt.initializer.name)
+              ? stmt.initializer.name.name
+              : undefined;
+
+            if (initializerName === undefined) {
+              throw new SynthError(
+                ErrorCodes.Unexpected_Error,
+                "The left-hand side of a 'for...in' statement cannot be a destructuring pattern."
+              );
+            }
 
-        if (initializerName === undefined) {
-          throw new SynthError(
-            ErrorCodes.Unsupported_Feature,
-            "Destructured parameter declarations are not yet supported by Step Functions. https://github.com/functionless/functionless/issues/364"
-          );
-        }
+            return {
+              startState: "assignIndex",
+              node: stmt.initializer,
+              states: {
+                assignIndex: {
+                  Type: "Pass",
+                  InputPath: `${tempArrayPath}[0].index`,
```

Comment:
What is `tempArrayPath`?

Reply:
We are doing tail recursion on the input array, which requires us to mutate an array variable. `tempArrayPath` is the copy of the original array we are mutating during the tail. (don't want to mutate the original array).